[dtensor] lazy init process groups in device mesh

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96700

This PR adds a private flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

wanchaol committed:
[dtensor] lazy init process groups in device mesh

This PR adds a flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

wanchaol committed:
Update on "[dtensor] lazy init process groups in device mesh"


This PR adds a flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] lazy init process groups in device mesh"


This PR adds a flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

wanchaol committed:
Update on "[dtensor] lazy init process groups in device mesh"


This PR adds a flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

wanchaol committed:
Update on "[dtensor] lazy init process groups in device mesh"


This PR adds a flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

fduwjj commented on torch/distributed/tensor/parallel/_utils.py:
```diff
@@ -149,4 +149,6 @@ def _create_1d_device_mesh(device_mesh: DeviceMesh, tp_mesh_dim: int = 0) -> Dev
     sub_pg = device_mesh.get_dim_groups()[tp_mesh_dim]
-    return DeviceMesh(device_mesh.device_type, dim_mesh_1d.squeeze(), [sub_pg])
+    mesh_1d = DeviceMesh(device_mesh.device_type, dim_mesh_1d.squeeze(), init_process_groups=False)
+    mesh_1d._dim_groups = [sub_pg]
```
Can we make override an API?

wanchaol replied:
Yeah I would prefer to do so if we really need this, but currently I think we should figure out how to slice a 2-D mesh to a 1-D mesh, if override is needed for the slice, we can definitely expose an API to do it

fduwjj commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -100,43 +100,88 @@ class DeviceMesh:
+        # TODO: we should do allgather the mesh tensor to ensure every rank have the same mesh value
+
+        # calculate the coordinates of the current global rank on the mesh
+        rank_coords = (self.mesh == get_rank()).nonzero()
```
For my learning, for 1D device mesh each rank is coordinate right?

XilunWu replied:
Assume the 1D device mesh (it can be a submesh) [1,3] on 4 ranks.
Because `nonzero()` returns a 2-D Tensor, so rank 1 has `rank_coords` be `[[0]]` and rank 3 has `[[1]]`.

wanchaol replied:
@fduwjj I think we should revisit how 1d device mesh is constructed from 2d mesh, the current way of doing it is somewhat weird. i.e. for 2-d mesh [[0, 1], [2, 3]], we should generate two 1-D mesh, instead of a single 1-D mesh.

fduwjj replied:
Since now we support sub-mesh. So essentially, we split one 2D mesh into two 1D sub-mesh. And we can decide which one to use based on the current rank. Is this something you are saying?

wanchaol replied:
Yes, so basically TP really only run in two sub meshes in a SPMD manner, i.e. in two hosts separately, it does not communicate across ranks(that's the parent mesh sync across the hosts).

fduwjj replied:
Do you think it makes sense to have a util func inside `device_mesh.py`  to split one 2D or N-D Mesh into lower dim submesh. Or we already have one?

wanchaol replied:
we don't have that yet. I think we can add one to TP first, we can think of whether to make this an API for device_mesh.py once we found there're more use cases :)

We should carefully maintain the device_mesh API going forward as it might be used internally too

fduwjj reviewed (approved):
LGTM

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -100,43 +100,88 @@ class DeviceMesh:
             if isinstance(mesh, torch.Tensor)
             else torch.tensor(mesh, dtype=torch.int)
         )
+        # always try to create default (world) pg, even if it is not initialized
```
@fduwjj I think it's as the comment said, we still need `self.rank` (e.g. in test code). We won't have it unless the world pg is created in `self._get_or_create_default_group()`.

wanchaol replied:
Yeah basically I think we at least need a world_pg in order to establish an rank identity

fduwjj replied:
I mean as long as we don't worry about the extra cost of calling `init_process_groups`, this should be fine.

wanchaol replied:
I am not particular sure if this is gonna be very useful, so making it a private arg for now. One thing I think that might be useful is that a non-initialized mesh can be used for sharding planning stage, so lazy init might be make sense.

XilunWu reviewed (approved):
lgtm but one question:
Does this PR  affect my default value PR?

wanchaol committed:
Update on "[dtensor] lazy init process groups in device mesh"


This PR adds a flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol committed:
rebase and make flag private on "[dtensor] lazy init process groups in device mesh"


This PR adds a flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] lazy init process groups in device mesh"


This PR adds a flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

wanchaol committed:
Update on "[dtensor] lazy init process groups in device mesh"


This PR adds a private flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] lazy init process groups in device mesh"


This PR adds a private flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

wanchaol commented:
> lgtm but one question: Does this PR affect my default value PR?

@XilunWu as discussed offline, this shouldn't affect the default value PR

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -f "ci all passed"
